### PR TITLE
Add code for synthetic preference pipeline

### PR DESCRIPTION
**Pull request by ljvmiranda921**
This PR adds a runbook and the source for running the synthetic data generation pipeline

### CODE REVIEW

File: scripts/synth_pref/README.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+First, let's generate configurations for `birr` (assuming your JSONL files in a directory called `source`):
+
+```sh
+python3 -m scripts.synth_pref.generate_responses \
+    --name myprompts \
+    --source_file "path/to/myprompts/*.jsonl" \
+    --target_dir "path/to/outputs/" \
+    --batch_size 128
+```
```
Comment (vwxyzjn) [inline scripts/synth_pref/README.md]: Can you include an example `jsonl`? Could be something dummy but would be good if the user can just copy-paste and run without doing any customizations. Once they run, they can customize their datasets.

File: scripts/synth_pref/README.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+This command will generate configuration files for each model that you can send to `birr`.
+To do so, run the following command:
+
+```sh
+python3 src/birr/batch_inference/runner.py --config-file path/to/config/file.yaml
```
Comment (vwxyzjn) [inline scripts/synth_pref/README.md:60]: Same comment as above, give a dummy yaml.

File: scripts/synth_pref/README.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+After running this command, you'll see all the outputs in the directory you specified for `--target_dir`.
+From there, you can create a preference annotation mix that samples four (4) responses for each model:
+
+```sh
+python3 -m scripts.synth_pref.create_annotation_mix \
+    --name myprompts \
+    --input_dir path/to/birr/output/ \
+    --output_dir path/to/annotation/output \
+    --prompt_template ultrafeedback
```
Comment (vwxyzjn) [inline scripts/synth_pref/README.md]: Same comment as above.

File: scripts/synth_pref/README.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+```
+
+Then, let's convert the annotation mix to the format desired by the Batch API:
+
+```sh
+python3 -m scripts.synth_pref.annotate_preferences \
+    --model gpt-4o-2024-08-06 \
+    --input_path path/to/the/annotation_mix.jsonl
+    --output_dir path/to/converted/output.jsonl
+    --rows_per_shard 10000
```
Comment (vwxyzjn) [inline scripts/synth_pref/README.md]: Same comment as above

File: scripts/synth_pref/README.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+```
+
+This command will shard our annotation mix to `n` files with `10000` rows to fit OpenAI's file limits.
+Then, we can send the annotations to OpenAI now:
+
+```sh
+python3 -m scripts.synth_pref.annotate_preferences \
+    --model gpt-4o-2024-08-06 \
+    --input_dir path/to/converted/files.jsonl
+    --output_dir target_dir
```
Comment (vwxyzjn) [inline scripts/synth_pref/README.md]: Same comment as above

File: scripts/synth_pref/README.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+```
+
+If all files are done and downloaded, you can start parsing the preferences to obtain the final preference dataset:
+
+```sh
+python3 -m scripts.synth_pref.parse_preferences \
+    --input_dir path/to/openai/outputs/ \
+    --output_path path/to/final/preference.yaml
+```
```
Comment (vwxyzjn) [inline scripts/synth_pref/README.md]: Same comment as above.

**Review by vwxyzjn — COMMENTED** [review]
Can we do something like `git clone birr && git checkout xxx_comit` instead of making it a submodule?

**Comment by ljvmiranda921** [comment]
@vwxyzjn updated! I'll resolve your comments

**Review by vwxyzjn — APPROVED** [review]
LGTM! Thanks

**Pull request merged**